### PR TITLE
fix(logger): use promise.then for processing query results(mysql)

### PR DESCRIPTION
**Pull request by Mikxail**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [ ] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

Using promise.then instead of custom handler function for processing run query results for mysql dialect.
It change fix losing CLS context for mysql logging custom function in case of `benchmark: true` #12317
MR for v5: #12318

### CODE REVIEW

**Comment by papb** [comment]
Hello! I see you are a first-time contributor, thank you for taking the time to help Sequelize! I hope to see more PRs from you in the future!

Please add at least one test to make sure your implementation works as intended and to prevent regressions in the future. You can write a test based on the example code you wrote in the issue. Take a look into recent merged PRs to see how others created tests. If you have trouble, let me know.

File: lib/dialects/mysql/query.js [hunk 1]
```diff
@@ -33,32 +33,25 @@ class Query extends AbstractQuery {
...
+      : new Promise((resolve, reject) => connection.query({ sql }, (error, result) => error ? reject(error) : resolve(result)).setMaxListeners(100));
 
-        if (err) {
-          // MySQL automatically rolls-back transactions in the event of a deadlock
-          if (options.transaction && err.errno === 1213) {
-            options.transaction.finished = 'rollback';
-          }
-          err.sql = sql;
-          err.parameters = parameters;
+    const complete = this._logQuery(sql, debug, parameters);
```
Comment (papb) [inline lib/dialects/mysql/query.js:40]: Please move this line above the query calls (as it was before your change).

File: lib/dialects/mysql/query.js [hunk 2]
```diff
@@ -33,32 +33,25 @@ class Query extends AbstractQuery {
...
+      if (options.transaction && err.errno === 1213) {
+        options.transaction.finished = 'rollback';
       }
+      err.sql = sql;
+      err.parameters = parameters;
+
+      throw this.formatError(err);
     });
+
+    complete();
```
Comment (papb) [inline lib/dialects/mysql/query.js:53]: Nice, I think it makes sense for `complete()` to be called only if no error happens. Good change :+1:

File: lib/dialects/mysql/query.js [hunk 3]
```diff
@@ -33,32 +33,25 @@ class Query extends AbstractQuery {
...
-        } else {
-          resolve(results);
-        }
-      };
-      if (parameters) {
-        debug('parameters(%j)', parameters);
-        connection.execute(sql, parameters, handler).setMaxListeners(100);
-      } else {
-        connection.query({ sql }, handler).setMaxListeners(100);
+    const results = await query.catch(err => {
```
Comment (papb) [inline lib/dialects/mysql/query.js:42]: Can you please refactor this `.catch` usage into a try-catch block?

**Comment by sushantdhiman** [comment]
Reminder this needs test before it can be reviewed :)

This issue must be affecting other dialects as well, having a test case will ensure that

File: lib/dialects/mysql/query.js (hunk 1 above)
Comment (Mikxail) [reply lib/dialects/mysql/query.js]: Done

File: lib/dialects/mysql/query.js (hunk 3 above)
Comment (Mikxail) [reply lib/dialects/mysql/query.js]: Sure. Done.

**Comment by Mikxail** [comment]
@papb Thank you for review. I made the requested changes. 
Also, I made changes in dialect/mssql for fixing same issue, but I hasn't idea why mssql test case brokes. Locally all tests passed.

File: lib/dialects/mssql/query.js [hunk 4]
```diff
@@ -112,6 +110,10 @@ class Query extends AbstractQuery {
 
       connection.execSql(request);
     });
+
+    complete();
```
Comment (sushantdhiman) [inline lib/dialects/mssql/query.js:114]: This means, benchmark will now include Sequelize record parsing overhead.

Earlier `benchmark = time to execute query by dialect`
Now `benchmark = time to execute query by dialect + time taken by sequelize to format result`

This deviation can be significant when parsing large number of records.

Main question is why placing `complete()` here works, do you have any ideas @Mikxail

Comment (Mikxail) [reply lib/dialects/mssql/query.js:114]: Actually, I don't know how `cls-hooked` works under the hood. But I can assume that the problem in `connection.lib.Request` that saves `callback` function in variable, and then call `callback` from `execSql`. Probably that can breaks `cls-hooked` algorithms.
A similar situation was in the `run` function (dialects/mysql/query.js).

I can refactor `_run` function (dialects/mssql/query.js), but it require a lot of changes. Can I do it withing this MR or I should create separate MR?

Comment (sushantdhiman) [reply lib/dialects/mssql/query.js:114]: I think issues listed below are variant of this issue
- https://github.com/nodejs/node/issues/21078
- https://github.com/Jeff-Lewis/cls-hooked/issues/36

At Sequelize level, we can extract record/error formatting away from callback. Something like this should work

1. Get raw result/error from callback where `triggerAsyncId` is incorrect (new Promise wrapper can be used)
2. Call `complete`
3. Parse error or raw results and return final result.

Comment (Mikxail) [reply lib/dialects/mssql/query.js]: I agree with it.
I did refactoring `_run` functions in dialects/mssql.

**Pull request merged**